File: services/data-pipeline/src/sale_procedure.py
from __future__ import annotations

import re
from datetime import UTC, datetime
from decimal import Decimal
from typing import Any

from src.models import AuctionSale
from src.normalize import clean_text

SALE_PROCEDURE_SCHEMA_VERSION = "sale_procedure_v1"
LEGAL_RULESET_VERSION = "fr_auction_participation_2026-08-20"

JUDICIAL_RULES_SOURCE_URL = "https://www.legifrance.gouv.fr/codes/section_lc/LEGITEXT000025024948/LEGISCTA000025939153/"
JUDICIAL_OVERBID_SOURCE_URL = (
    "https://www.legifrance.gouv.fr/codes/section_lc/LEGITEXT000025024948/LEGISCTA000025939177/"
)
NOTARIAL_RULES_SOURCE_URL = (
    "https://www.immobilier.notaires.fr/fr/articles/conseils-et-actualites/"
    "achat-vente/achat-maison-appartement-aux-encheres-achat-immobilier-encheres"
)

_EXPLICIT_TRIBUNAL_PATTERNS = (
    re.compile(r"\btribunal\s+judiciaire\b", re.I),
    re.compile(r"\bvente\s+(?:aux\s+ench[eè]res\s+)?(?:a|à)\s+la\s+barre\b", re.I),
    re.compile(r"\baudience\s+d['’]?adjudication\b", re.I),
    re.compile(r"\bavocat\s+inscrit\s+au\s+barreau\b", re.I),
)
_EXPLICIT_NOTARY_PATTERNS = (
    re.compile(r"\bvente\s+(?:aux\s+ench[eè]res\s+)?notariale\b", re.I),
    re.compile(r"\badjudication\s+(?:devant|par)\s+(?:un\s+)?notaire\b", re.I),
    re.compile(r"\bchambre\s+(?:interd[eé]partementale\s+|d[eé]partementale\s+)?des\s+notaires\b", re.I),
    re.compile(r"\bimmo[-\s]?interactif\b", re.I),
)
_STATE_PATTERNS = (
    re.compile(r"\bvente\s+domaniale\b", re.I),
    re.compile(r"\bcessions?\s+immobili[eè]res?\s+de\s+l['’][eé]tat\b", re.I),
    re.compile(r"\bdirection\s+de\s+l['’]immobilier\s+de\s+l['’][eé]tat\b", re.I),
)


def classify_sale_procedure(
    sale: AuctionSale,
    *,
    verified_at: str | None = None,
) -> AuctionSale:
    """Attach a conservative, evidence-carrying participation profile to a sale.

    The classifier distinguishes the venue/operator from the legal framework.
    It never upgrades an address-only court inference to a verified sale venue.
    """

    checked_at = verified_at or datetime.now(UTC).isoformat()
    corpus = _sale_corpus(sale)
    judicial_matches = _matched_signals(corpus, _EXPLICIT_TRIBUNAL_PATTERNS)
    notarial_matches = _matched_signals(corpus, _EXPLICIT_NOTARY_PATTERNS)
    state_matches = _matched_signals(corpus, _STATE_PATTERNS)

    if sale.source_name == "notaires":
        notarial_matches.append("Source officielle des Notaires de France")
    if sale.source_name == "cessions_etat":
        state_matches.append("Source officielle des cessions immobilières de l'État")

    venue_type, status, issues = _resolve_venue(
        sale,
        judicial_matches=judicial_matches,
        notarial_matches=notarial_matches,
        state_matches=state_matches,
    )
    rules_venue_type = venue_type if status in {"verified", "cross_checked"} else "unknown"
    legal_framework = _resolve_legal_framework(corpus, venue_type)
    participation_mode = _participation_mode(corpus, rules_venue_type)
    case_sources = _case_sources(sale)
    rules, regulatory_sources = _participation_rules(
        sale,
        venue_type=rules_venue_type,
        corpus=corpus,
        checked_at=checked_at,
    )
    facts = _verification_facts(
        sale,
        venue_type=venue_type,
        legal_framework=legal_framework,
        status=status,
        judicial_matches=judicial_matches,
        notarial_matches=notarial_matches,
        state_matches=state_matches,
    )

    procedure = {
        "schema_version": SALE_PROCEDURE_SCHEMA_VERSION,
        "ruleset_version": LEGAL_RULESET_VERSION,
        "venue_type": venue_type,
        "legal_framework": legal_framework,
        "venue_name": _venue_name(sale, venue_type),
        "venue_address": _venue_address(sale),
        "participation_mode": participation_mode,
        "organizer_name": _organizer_name(sale, venue_type),
        "organizer_type": _organizer_type(venue_type),
        "organizer_contact": clean_text(sale.lawyer_contact),
        "eligible_bar": _eligible_bar(sale, venue_type),
        "rules": rules,
        "verification": {
            "status": status,
            "verified_at": checked_at,
            "case_source_count": len(case_sources),
            "case_sources": case_sources,
            "regulatory_sources": regulatory_sources,
            "facts": facts,
            "issues": issues,
        },
    }

    sale.sale_venue_type = venue_type
    sale.sale_legal_framework = legal_framework
    sale.sale_verification_status = status
    sale.sale_procedure = procedure
    sale.raw_payload["sale_procedure"] = procedure
    source_blocks = sale.raw_payload.setdefault("source_blocks", {})
    if isinstance(source_blocks, dict):
        source_blocks["sale_procedure"] = procedure
    _replace_quality_flag(sale, "sale_procedure_unverified", status == "pending")
    _replace_quality_flag(sale, "sale_procedure_conflict", status == "conflict")
    return sale


def _resolve_venue(
    sale: AuctionSale,
    *,
    judicial_matches: list[str],
    notarial_matches: list[str],
    state_matches: list[str],
) -> tuple[str, str, list[str]]:
    issues: list[str] = []
    explicit_families = sum(bool(value) for value in (judicial_matches, notarial_matches, state_matches))
    if explicit_families > 1:
        issues.append("Des indices contradictoires de lieu ou d'organisateur doivent être relus.")
        return "unknown", "conflict", issues

    if judicial_matches:
        has_official_mapping = _has_verified_court_assignment(sale)
        return "tribunal", "cross_checked" if has_official_mapping else "verified", issues
    if notarial_matches:
        return "notary", "cross_checked" if sale.source_name == "notaires" else "verified", issues
    if state_matches:
        return "state", "cross_checked" if sale.source_name == "cessions_etat" else "verified", issues

    if sale.tribunal or sale.tribunal_code:
        issues.append(
            "Le tribunal compétent est connu par l'adresse, mais le lieu de cette vente n'est pas encore confirmé par l'annonce."
        )
        return "tribunal", "pending", issues
    if sale.source_name == "notaires":
        issues.append("Le canal notarial est déduit de la source, sans mention de lieu suffisamment explicite.")
        return "notary", "pending", issues
    if sale.source_name == "cessions_etat":
        issues.append("Le canal domanial est déduit de la source, sans modalités particulières confirmées.")
        return "state", "pending", issues

    issues.append("Le lieu et le mode de participation ne sont pas encore confirmés.")
    return "unknown", "pending", issues


def _participation_rules(
    sale: AuctionSale,
    *,
    venue_type: str,
    corpus: str,
    checked_at: str,
) -> tuple[dict[str, Any], list[dict[str, Any]]]:
    if venue_type == "tribunal":
        guarantee = _judicial_guarantee(sale.starting_price_eur)
        return (
            {
                "lawyer_required": True,
                "lawyer_note": (
                    "Les enchères sont portées par un avocat inscrit au barreau du tribunal "
                    "judiciaire devant lequel la vente est poursuivie."
                ),
                "bid_method": "lawyer_mandate",
                "guarantee": {
                    "amount_eur": guarantee,
                    "rate_pct": 10,
                    "minimum_eur": 3000,
                    "status": "regulatory_verified",
                    "note": "Garantie légale minimale ; le bénéficiaire et les autres chèques sont précisés par le cahier des conditions de vente.",
                },
                "financing_condition": False,
                "cooling_off_period": False,
                "payment_deadline_days": 60,
                "overbid": {
                    "allowed": True,
                    "minimum_increase_pct": 10,
                    "window_days": 10,
                    "note": "La surenchère est formée par acte d'avocat.",
                },
            },
            [
                _regulatory_source(
                    "Code des procédures civiles d'exécution — enchères et garantie",
                    JUDICIAL_RULES_SOURCE_URL,
                    checked_at,
                ),
                _regulatory_source(
                    "Code des procédures civiles d'exécution — surenchère",
                    JUDICIAL_OVERBID_SOURCE_URL,
                    checked_at,
                ),
            ],
        )

    if venue_type == "notary":
        amount_eur, rate_pct = _extract_notarial_guarantee(corpus)
        return (
            {
                "lawyer_required": False,
                "lawyer_note": (
                    "L'acquéreur peut en principe porter lui-même les enchères ; les modalités "
                    "particulières du cahier des charges restent prioritaires."
                ),
                "bid_method": "direct_or_sale_specific",
                "guarantee": {
                    "amount_eur": amount_eur,
                    "rate_pct": rate_pct,
                    "minimum_eur": None,
                    "status": "case_verified"
                    if amount_eur is not None or rate_pct is not None
                    else "pending_case_document",
                    "note": (
                        "Montant relevé dans les informations de la vente."
                        if amount_eur is not None or rate_pct is not None
                        else "Le montant exact doit être confirmé dans le cahier des charges."
                    ),
                },
                "financing_condition": False,
                "cooling_off_period": False,
                "payment_deadline_days": 45,
                "overbid": {
                    "allowed": None,
                    "minimum_increase_pct": None,
                    "window_days": None,
                    "note": "La faculté de surenchère dépend du cahier des charges de la vente.",
                },
            },
            [
                _regulatory_source(
                    "Guide des ventes aux enchères des Notaires de France",
                    NOTARIAL_RULES_SOURCE_URL,
                    checked_at,
                )
            ],
        )

    return (
        {
            "lawyer_required": None,
            "lawyer_note": "Les modalités particulières de cette vente restent à confirmer.",
            "bid_method": "sale_specific",
            "guarantee": {
                "amount_eur": None,
                "rate_pct": None,
                "minimum_eur": None,
                "status": "pending_case_document",
                "note": "Consignation à confirmer dans les conditions de vente.",
            },
            "financing_condition": None,
            "cooling_off_period": None,
            "payment_deadline_days": None,
            "overbid": {
                "allowed": None,
                "minimum_increase_pct": None,
                "window_days": None,
                "note": "Règle propre à la vente à confirmer.",
            },
        },
        [],
    )


def _verification_facts(
    sale: AuctionSale,
    *,
    venue_type: str,
    legal_framework: str,
    status: str,
    judicial_matches: list[str],
    notarial_matches: list[str],
    state_matches: list[str],
) -> list[dict[str, Any]]:
    matches = judicial_matches or notarial_matches or state_matches
    facts = [
        {
            "key": "venue_type",
            "value": venue_type,
            "status": status,
            "evidence": matches[:4],
            "source_url": sale.source_url,
        },
        {
            "key": "legal_framework",
            "value": legal_framework,
            "status": "verified" if legal_framework != "unknown" and matches else "pending",
            "evidence": _framework_signals(legal_framework),
            "source_url": sale.source_url,
        },
    ]
    if _has_verified_court_assignment(sale):
        assignment = sale.raw_payload["tribunal_assignment"]
        facts.append(
            {
                "key": "competent_court",
                "value": assignment.get("court_name"),
                "status": "verified",
                "evidence": [
                    f"Commune INSEE {assignment.get('insee_code')}",
                    str(assignment.get("mapping_method") or ""),
                ],
                "source_url": assignment.get("source_url"),
            }
        )
    return facts


def _sale_corpus(sale: AuctionSale) -> str:
    payload = sale.raw_payload if isinstance(sale.raw_payload, dict) else {}
    source_blocks = payload.get("source_blocks")
    # Derived procedure output must never become evidence for its next run.
    # Otherwise a pending address-only court assignment can promote itself to
    # verified by matching wording in its own generated participation guide.
    block_values = (
        [value for key, value in source_blocks.items() if key != "sale_procedure"]
        if isinstance(source_blocks, dict)
        else []
    )
    document_values: list[object] = []
    for document in sale.documents:
        if not isinstance(document, dict):
            continue
        document_values.extend(document.get(key) for key in ("name", "label", "type", "document_type"))
    values: list[object] = [
        sale.title,
        sale.description,
        sale.raw_text,
        sale.source_name,
        sale.primary_source,
        payload.get("source_description"),
        *block_values,
        *document_values,
    ]
    return clean_text("\n".join(str(value) for value in values if value)) or ""


def _matched_signals(corpus: str, patterns: tuple[re.Pattern[str], ...]) -> list[str]:
    matches: list[str] = []
    for pattern in patterns:
        match = pattern.search(corpus)
        if match:
            value = clean_text(match.group(0))
            if value and value not in matches:
                matches.append(value)
    return matches


def _resolve_legal_framework(corpus: str, venue_type: str) -> str:
    patterns = (
        ("judicial_seizure", r"\bsaisie\s+immobili[eè]re\b"),
        ("insolvency", r"\bliquidation\s+judiciaire\b|\bproc[eé]dure\s+collective\b"),
        ("judicial_partition", r"\blicitation\b|\bpartage\s+judiciaire\b"),
        ("state_sale", r"\bvente\s+domaniale\b|\bcessions?\s+immobili[eè]res?\s+de\s+l['’][eé]tat\b"),
        ("voluntary_notarial", r"\bvente\s+volontaire\b|\bvente\s+notariale\b"),
    )
    for value, pattern in patterns:
        if re.search(pattern, corpus, re.I):
            return value
    if venue_type == "state":
        return "state_sale"
    return "unknown"


def _framework_signals(framework: str) -> list[str]:
    labels = {
        "judicial_seizure": "Mention de saisie immobilière",
        "insolvency": "Mention de liquidation ou procédure collective",
        "judicial_partition": "Mention de licitation ou partage judiciaire",
        "state_sale": "Mention de vente domaniale",
        "voluntary_notarial": "Mention de vente volontaire ou notariale",
    }
    return [labels[framework]] if framework in labels else []


def _case_sources(sale: AuctionSale) -> list[dict[str, Any]]:
    sources: list[dict[str, Any]] = [
        {
            "kind": "listing",
            "label": clean_text(sale.primary_source) or clean_text(sale.source_name) or "Annonce",
            "source_name": clean_text(sale.source_name) or "Annonce",
            "url": sale.source_url,
        }
    ]
    seen = {sale.source_url}
    for document in sale.documents:
        if not isinstance(document, dict):
            continue
        url = clean_text(document.get("url") or document.get("document_url"))
        if not url or url in seen:
            continue
        seen.add(url)
        sources.append(
            {
                "kind": "document",
                "label": clean_text(document.get("name") or document.get("label")) or "Pièce du dossier",
                "source_name": clean_text(sale.source_name) or "Dossier de vente",
                "url": url,
                "document_type": clean_text(document.get("document_type") or document.get("type")),
            }
        )
        if len(sources) >= 6:
            break
    assignment = sale.raw_payload.get("tribunal_assignment")
    if _has_verified_court_assignment(sale) and isinstance(assignment, dict):
        url = clean_text(assignment.get("source_url"))
        if url and url not in seen:
            sources.append(
                {
                    "kind": "official_reference",
                    "label": "Référentiel officiel de compétence territoriale",
                    "source_name": "Ministère de la Justice",
                    "url": url,
                    "reference_sha256": assignment.get("reference_sha256"),
                }
            )
    return sources


def _has_verified_court_assignment(sale: AuctionSale) -> bool:
    assignment = sale.raw_payload.get("tribunal_assignment")
    return bool(
        isinstance(assignment, dict)
        and assignment.get("status") == "verified"
        and assignment.get("mapping_method") == "justice_competence_insee_exact"
        and assignment.get("court_code")
        and assignment.get("source_url")
    )


def _venue_name(sale: AuctionSale, venue_type: str) -> str | None:
    if venue_type == "tribunal":
        return clean_text(sale.tribunal) or clean_text(sale.tribunal_code)
    if venue_type == "notary":
        return _extract_notary_name(_sale_corpus(sale)) or clean_text(sale.lawyer_name)
    if venue_type == "state":
        return "Vente immobilière de l'État"
    return None


def _venue_address(sale: AuctionSale) -> str | None:
    payload = sale.raw_payload if isinstance(sale.raw_payload, dict) else {}
    source_blocks = payload.get("source_blocks")
    if isinstance(source_blocks, dict):
        for key in ("sale_location", "auction_location", "venue", "lieu_vente"):
            value = clean_text(source_blocks.get(key))
            if value:
                return value
    return None


def _organizer_name(sale: AuctionSale, venue_type: str) -> str | None:
    if venue_type in {"tribunal", "notary"}:
        return clean_text(sale.lawyer_name) or (
            _extract_notary_name(_sale_corpus(sale)) if venue_type == "notary" else None
        )
    if venue_type == "state":
        return "Direction de l'immobilier de l'État"
    return None


def _organizer_type(venue_type: str) -> str:
    return {
        "tribunal": "pursuing_lawyer",
        "notary": "notary",
        "state": "state_service",
    }.get(venue_type, "unknown")


def _eligible_bar(sale: AuctionSale, venue_type: str) -> str | None:
    if venue_type != "tribunal":
        return None
    city = clean_text(sale.raw_payload.get("tribunal_assignment", {}).get("court_city"))
    if not city:
        city = clean_text(sale.tribunal)
        city = re.sub(r"^TJ\s+", "", city or "", flags=re.I) or None
    return f"Barreau de {city}" if city else None


def _participation_mode(corpus: str, venue_type: str) -> str:
    online = bool(re.search(r"\ben\s+ligne\b|\bimmo[-\s]?interactif\b|\bvisioconf[eé]rence\b", corpus, re.I))
    in_person = bool(
        re.search(
            r"\ben\s+salle\b|\bau\s+tribunal\b|\bpalais\s+de\s+justice\b|\bchambre\s+des\s+notaires\b", corpus, re.I
        )
    )
    if online and in_person:
        return "hybrid"
    if online:
        return "online"
    if in_person or venue_type == "tribunal":
        return "in_person"
    return "unknown"


def _judicial_guarantee(starting_price: Decimal | None) -> float | None:
    if starting_price is None:
        return None
    return float(max(Decimal("3000"), starting_price * Decimal("0.10")))


def _extract_notarial_guarantee(corpus: str) -> tuple[float | None, float | None]:
    amount_match = re.search(
        r"\bconsignation\b.{0,100}?([0-9][0-9\s.]{2,})\s*(?:€|euros?)\b",
        corpus,
        re.I,
    )
    rate_match = re.search(r"\bconsignation\b.{0,80}?(\d{1,2}(?:[,.]\d+)?)\s*%", corpus, re.I)
    amount = _parse_amount(amount_match.group(1)) if amount_match else None
    rate = float(rate_match.group(1).replace(",", ".")) if rate_match else None
    return amount, rate


def _parse_amount(value: str) -> float | None:
    cleaned = re.sub(r"[^0-9]", "", value)
    return float(cleaned) if cleaned else None


def _extract_notary_name(corpus: str) -> str | None:
    match = re.search(
        r"\b(?:Ma[iî]tre|Me)\s+([A-ZÀ-ÖØ-Ý][A-Za-zÀ-ÿ'’ -]{2,60})",
        corpus,
    )
    return f"Me {clean_text(match.group(1))}" if match else None


def _regulatory_source(label: str, url: str, checked_at: str) -> dict[str, str]:
    return {
        "kind": "legal_basis",
        "label": label,
        "source_name": "Source officielle",
        "url": url,
        "checked_at": checked_at,
        "ruleset_version": LEGAL_RULESET_VERSION,
    }


def _replace_quality_flag(sale: AuctionSale, flag: str, enabled: bool) -> None:
    sale.quality_flags = [value for value in sale.quality_flags if value != flag]
    if enabled:
        sale.quality_flags.append(flag)

File: src/lib/sale-procedure.ts
import { z } from "zod";
import { formatPrice } from "@/lib/format";
import type {
  AuctionSale,
  SaleLegalFramework,
  SaleVenueType,
  SaleVerificationStatus,
} from "@/lib/types";

const venueTypeSchema = z.enum(["tribunal", "notary", "state", "online", "unknown"]);
const legalFrameworkSchema = z.enum([
  "judicial_seizure",
  "judicial_partition",
  "insolvency",
  "voluntary_notarial",
  "state_sale",
  "unknown",
]);
const verificationStatusSchema = z.enum(["verified", "cross_checked", "pending", "conflict"]);
const nullableNumber = z.number().finite().nonnegative().nullable();

const sourceSchema = z
  .object({
    kind: z.string().optional(),
    label: z.string().optional(),
    source_name: z.string().optional(),
    url: z.string().nullable().optional(),
    checked_at: z.string().nullable().optional(),
    document_type: z.string().nullable().optional(),
  })
  .passthrough();

const factSchema = z
  .object({
    key: z.string(),
    value: z.unknown(),
    status: verificationStatusSchema,
    evidence: z.array(z.string()).default([]),
    source_url: z.string().nullable().optional(),
  })
  .passthrough();

export const saleProcedureSchema = z
  .object({
    schema_version: z.literal("sale_procedure_v1"),
    ruleset_version: z.string().min(1),
    venue_type: venueTypeSchema,
    legal_framework: legalFrameworkSchema,
    venue_name: z.string().nullable(),
    venue_address: z.string().nullable(),
    participation_mode: z.enum(["in_person", "online", "hybrid", "unknown"]),
    organizer_name: z.string().nullable(),
    organizer_type: z.string(),
    organizer_contact: z.string().nullable(),
    eligible_bar: z.string().nullable(),
    rules: z
      .object({
        lawyer_required: z.boolean().nullable(),
        lawyer_note: z.string(),
        bid_method: z.string(),
        guarantee: z
          .object({
            amount_eur: nullableNumber,
            rate_pct: nullableNumber,
            minimum_eur: nullableNumber,
            status: z.string(),
            note: z.string(),
          })
          .passthrough(),
        financing_condition: z.boolean().nullable(),
        cooling_off_period: z.boolean().nullable(),
        payment_deadline_days: z.number().int().positive().nullable(),
        overbid: z
          .object({
            allowed: z.boolean().nullable(),
            minimum_increase_pct: nullableNumber,
            window_days: z.number().int().positive().nullable(),
            note: z.string(),
          })
          .passthrough(),
      })
      .passthrough(),
    verification: z
      .object({
        status: verificationStatusSchema,
        verified_at: z.string().nullable(),
        case_source_count: z.number().int().nonnegative(),
        case_sources: z.array(sourceSchema),
        regulatory_sources: z.array(sourceSchema),
        facts: z.array(factSchema),
        issues: z.array(z.string()),
      })
      .passthrough(),
  })
  .passthrough();

export type SaleProcedure = z.infer<typeof saleProcedureSchema>;
export type SaleProcedureSource = z.infer<typeof sourceSchema>;

export type SaleProcedurePresentation = {
  procedure: SaleProcedure | null;
  venueType: SaleVenueType;
  legalFramework: SaleLegalFramework;
  verificationStatus: SaleVerificationStatus;
  verifiedAt: string | null;
  venueName: string | null;
  venueAddress: string | null;
  organizerName: string | null;
  organizerContact: string | null;
  eligibleBar: string | null;
  participationMode: "in_person" | "online" | "hybrid" | "unknown";
  lawyerRequired: boolean | null;
  lawyerNote: string;
  guaranteeAmountEur: number | null;
  guaranteeRatePct: number | null;
  guaranteeMinimumEur: number | null;
  guaranteeNote: string;
  financingCondition: boolean | null;
  coolingOffPeriod: boolean | null;
  paymentDeadlineDays: number | null;
  overbidAllowed: boolean | null;
  overbidMinimumIncreasePct: number | null;
  overbidWindowDays: number | null;
  overbidNote: string;
  sources: SaleProcedureSource[];
  issues: string[];
};

export function getSaleProcedure(sale: AuctionSale): SaleProcedurePresentation {
  const embedded = embeddedProcedure(sale);
  const parsed = saleProcedureSchema.safeParse(embedded);
  const procedure = parsed.success ? parsed.data : null;
  const venueType =
    procedure?.venue_type ?? parseVenueType(sale.sale_venue_type) ?? legacyVenueType(sale);
  const legalFramework =
    procedure?.legal_framework ?? parseLegalFramework(sale.sale_legal_framework) ?? "unknown";
  const verificationStatus =
    procedure?.verification.status ??
    parseVerificationStatus(sale.sale_verification_status) ??
    "pending";
  const venueIsConfirmed =
    verificationStatus === "verified" || verificationStatus === "cross_checked";
  const rules = procedure?.rules;

  return {
    procedure,
    venueType,
    legalFramework,
    verificationStatus,
    verifiedAt: procedure?.verification.verified_at ?? null,
    venueName:
      procedure?.venue_name ??
      (venueType === "tribunal"
        ? (sale.tribunal_name ?? sale.tribunal)
        : venueType === "notary"
          ? sale.lawyer_name
          : null),
    venueAddress: procedure?.venue_address ?? null,
    organizerName: procedure?.organizer_name ?? sale.lawyer_name,
    organizerContact: procedure?.organizer_contact ?? sale.lawyer_contact,
    eligibleBar:
      procedure?.eligible_bar ?? (venueType === "tribunal" ? legacyEligibleBar(sale) : null),
    participationMode:
      procedure?.participation_mode ??
      (venueType === "tribunal" && venueIsConfirmed ? "in_person" : "unknown"),
    lawyerRequired:
      rules?.lawyer_required ?? (venueType === "tribunal" && venueIsConfirmed ? true : null),
    lawyerNote:
      rules?.lawyer_note ??
      (venueType === "tribunal" && venueIsConfirmed
        ? "Les enchères sont portées par un avocat du barreau compétent. Cette règle doit encore être rapprochée des pièces de la vente."
        : "Les modalités de représentation doivent être confirmées dans les conditions de vente."),
    guaranteeAmountEur:
      rules?.guarantee.amount_eur ??
      (venueType === "tribunal" && venueIsConfirmed
        ? judicialGuarantee(sale.starting_price_eur)
        : null),
    guaranteeRatePct:
      rules?.guarantee.rate_pct ?? (venueType === "tribunal" && venueIsConfirmed ? 10 : null),
    guaranteeMinimumEur:
      rules?.guarantee.minimum_eur ?? (venueType === "tribunal" && venueIsConfirmed ? 3000 : null),
    guaranteeNote:
      rules?.guarantee.note ??
      "Le montant exact et le bénéficiaire doivent être confirmés dans les conditions de vente.",
    financingCondition: rules?.financing_condition ?? null,
    coolingOffPeriod: rules?.cooling_off_period ?? null,
    paymentDeadlineDays: rules?.payment_deadline_days ?? null,
    overbidAllowed: rules?.overbid.allowed ?? null,
    overbidMinimumIncreasePct: rules?.overbid.minimum_increase_pct ?? null,
    overbidWindowDays: rules?.overbid.window_days ?? null,
    overbidNote:
      rules?.overbid.note ?? "La faculté de surenchère doit être confirmée pour cette vente.",
    sources: procedure
      ? [...procedure.verification.case_sources, ...procedure.verification.regulatory_sources]
      : [],
    issues: procedure?.verification.issues ?? ["Qualification détaillée en cours de vérification."],
  };
}

export function saleVenueLabel(venueType: SaleVenueType): string {
  return {
    tribunal: "Vente au tribunal",
    notary: "Vente notariale",
    state: "Vente domaniale",
    online: "Vente en ligne",
    unknown: "Mode de vente à confirmer",
  }[venueType];
}

export function saleVenueShortLabel(venueType: SaleVenueType): string {
  return {
    tribunal: "Tribunal",
    notary: "Notaire",
    state: "État",
    online: "En ligne",
    unknown: "À confirmer",
  }[venueType];
}

export function saleVerificationLabel(status: SaleVerificationStatus): string {
  return {
    verified: "Vérifié",
    cross_checked: "Recoupé",
    pending: "En cours de vérification",
    conflict: "Contrôle requis",
  }[status];
}

export function saleProcedureIsConfirmed(procedure: SaleProcedurePresentation): boolean {
  return (
    procedure.verificationStatus === "verified" || procedure.verificationStatus === "cross_checked"
  );
}

export function saleIsTribunalVenue(sale: AuctionSale): boolean {
  return getSaleProcedure(sale).venueType === "tribunal";
}

export function saleLegalFrameworkLabel(framework: SaleLegalFramework): string {
  return {
    judicial_seizure: "Saisie immobilière",
    judicial_partition: "Licitation ou partage judiciaire",
    insolvency: "Liquidation ou procédure collective",
    voluntary_notarial: "Vente notariale volontaire",
    state_sale: "Vente domaniale",
    unknown: "Cadre juridique à confirmer",
  }[framework];
}

export function participationModeLabel(
  value: SaleProcedurePresentation["participationMode"],
): string {
  return {
    in_person: "Sur place",
    online: "En ligne",
    hybrid: "Sur place et en ligne",
    unknown: "À confirmer",
  }[value];
}

export function lawyerRequirementLabel(procedure: SaleProcedurePresentation): string {
  if (procedure.lawyerRequired === true) return "Avocat obligatoire pour enchérir";
  if (procedure.lawyerRequired === false) return "Avocat non obligatoire en principe";
  return "Représentation à confirmer";
}

export function guaranteeLabel(procedure: SaleProcedurePresentation): string {
  const amount = procedure.guaranteeAmountEur;
  const rate = procedure.guaranteeRatePct;
  const minimum = procedure.guaranteeMinimumEur;
  if (amount != null) {
    const suffix = rate != null ? ` · ${formatPercentNumber(rate)} de la mise à prix` : "";
    return `${formatPrice(amount)}${suffix}`;
  }
  if (rate != null) {
    return `${formatPercentNumber(rate)} de la mise à prix${
      minimum != null ? ` · minimum ${formatPrice(minimum)}` : ""
    }`;
  }
  return "Montant à confirmer";
}

export function paymentDeadlineLabel(days: number | null): string {
  if (days == null) return "À confirmer dans les conditions de vente";
  if (days === 60) return "2 mois après l’adjudication définitive";
  return `${days} jours après l’adjudication`;
}

export function overbidLabel(procedure: SaleProcedurePresentation): string {
  if (procedure.overbidAllowed === false) return "Non prévue";
  if (procedure.overbidAllowed !== true) return "À confirmer dans les conditions de vente";
  const parts = [
    procedure.overbidMinimumIncreasePct != null
      ? `+${formatPercentNumber(procedure.overbidMinimumIncreasePct)} minimum`
      : null,
    procedure.overbidWindowDays != null ? `dans les ${procedure.overbidWindowDays} jours` : null,
  ].filter(Boolean);
  return parts.join(" · ") || "Possible";
}

export function saleEventLabel(venueType: SaleVenueType): string {
  return venueType === "tribunal" ? "Audience" : "Vente";
}

function embeddedProcedure(sale: AuctionSale): unknown {
  if (sale.sale_procedure) return sale.sale_procedure;
  const sourceBlocks = sale.source_blocks;
  if (sourceBlocks && typeof sourceBlocks === "object") {
    return sourceBlocks.sale_procedure;
  }
  return null;
}

function legacyVenueType(sale: AuctionSale): SaleVenueType {
  const source = `${sale.source_name ?? ""} ${sale.primary_source ?? ""}`.toLowerCase();
  if (source.includes("notaire")) return "notary";
  if (source.includes("cession") || source.includes("domanial")) return "state";
  if (sale.tribunal || sale.tribunal_code || sale.tribunal_name) return "tribunal";
  return "unknown";
}

function parseVenueType(value: unknown): SaleVenueType | null {
  const parsed = venueTypeSchema.safeParse(value);
  return parsed.success ? parsed.data : null;
}

function parseLegalFramework(value: unknown): SaleLegalFramework | null {
  const parsed = legalFrameworkSchema.safeParse(value);
  return parsed.success ? parsed.data : null;
}

function parseVerificationStatus(value: unknown): SaleVerificationStatus | null {
  const parsed = verificationStatusSchema.safeParse(value);
  return parsed.success ? parsed.data : null;
}

function legacyEligibleBar(sale: AuctionSale): string | null {
  const city = sale.tribunal_city ?? sale.tribunal_name?.replace(/^TJ\s+/i, "") ?? null;
  return city ? `Barreau de ${city}` : null;
}

function judicialGuarantee(startingPrice: number | null): number | null {
  if (startingPrice == null || !Number.isFinite(startingPrice)) return null;
  return Math.max(3000, startingPrice * 0.1);
}

function formatPercentNumber(value: number): string {
  return `${new Intl.NumberFormat("fr-FR", { maximumFractionDigits: 1 }).format(value)} %`;
}
